File: windows/CodexAccountSwitcher.Core/AccountSnapshot.cs
using System.Text.Json;

namespace CodexAccountSwitcher.Core;

// Wire DTOs only. Account policy and storage live in Sources/SwitcherCore.
public sealed record Profile(Guid Id, string DisplayName, string? Email);
public sealed record Usage(int RemainingPercent, DateTimeOffset ResetsAt,
    int? FiveHourRemainingPercent, DateTimeOffset? FiveHourResetsAt);
public sealed record AccountRow(Profile Profile, string Initials, Usage? Usage, string? UsageError,
    string UsageStatus, bool IsActive, bool IsCredentialOwner, bool CanRemove);
public sealed record ProviderProfile(string Id, string DisplayName);
public sealed record ProviderRow(ProviderProfile Profile, bool IsActive, string Subtitle);
public sealed record SwitchConfirmation(Guid? AccountID, string? ProviderID, string Title, string Message, string ConfirmTitle);
public sealed record ProviderModel(string Id, string DisplayName, bool IsEnabled, string[] ReasoningOptions, string? ReasoningEffort);
public sealed record ManagedProvider(string Id, string DisplayName, string BaseURL, string ApiFormat, ProviderModel[] Models, string DefaultModelID, string Sort = "custom");
public sealed record ProviderEditorState(string Id, string DisplayName, string BaseURL, string ApiFormat, bool HasStoredKey,
    ProviderModel[] Models, string? DefaultModelID, string Query, string Sort, string[] VisibleModelIDs, bool IsBusy, string? Error, bool DidSave);
public sealed record ProviderConnectionInput(string DisplayName, string BaseURL, string ApiFormat, string? ApiKey);
public sealed record ProviderEditorCommand(string? ProviderID = null, ProviderConnectionInput? Connection = null,
    string? ModelID = null, string? Query = null, string? Sort = null, int? Offset = null, bool? Value = null, string? Effort = null);
public sealed record Preferences(string Language = "system", bool ShowsMenuBarPercentage = true,
    bool ShowsFiveHourUsage = false, bool EnablesProviderSwitching = false);
public sealed record AccountSnapshot(AccountRow[] Accounts, Guid? ActiveAccountID, Preferences Settings,
    bool IsMutating, bool IsAddingAccount, bool ActiveIdentityConfirmed, string? Error, Dictionary<string, string> Strings,
    ProviderRow[] Providers, string ActiveProviderID, string AuthenticationKind, SwitchConfirmation? PendingSwitch,
    ManagedProvider[] ManagedProviders, ProviderEditorState? ProviderEditor)
{
    public static AccountSnapshot Empty { get; } = new([], null, new(), false, false, false, null, [], [], "openai", "unknown", null, [], null);
    public string Text(string key) => Strings.GetValueOrDefault(key, key);
    public static JsonSerializerOptions JsonOptions { get; } = new() { PropertyNameCaseInsensitive = true };
}

public interface IAccountClient
{
    AccountSnapshot State { get; }
    event Action? Changed;
    Task CommandAsync(string command, Guid? accountID = null, bool? value = null, string? language = null, string? providerID = null);
    Task ProviderCommandAsync(string command, ProviderEditorCommand? editor = null);
}

File: Sources/SwitcherCore/AccountSnapshot.swift
import Foundation

/// Presentation data only. Credentials and authentication tokens never leave the core.
public struct AccountSnapshot: Encodable, Sendable {
    public struct Row: Encodable, Sendable {
        public let profile: AccountProfile
        public let initials: String
        public let usage: WeeklyUsage?
        public let usageError: String?
        public let usageStatus: String
        public let isActive: Bool
        public let isCredentialOwner: Bool
        public let canRemove: Bool
    }
    public struct ProviderRow: Encodable, Sendable {
        public let profile: ProviderProfile
        public let isActive: Bool
        public let subtitle: String
    }
    public let accounts: [Row]
    public let providers: [ProviderRow]
    public let activeProviderID: String
    public let authenticationKind: String
    public let pendingSwitch: SwitchConfirmation?
    public let managedProviders: [ManagedProvider]
    public let providerEditor: ProviderEditorState?
    public let activeAccountID: UUID?
    public let settings: AppSettings
    public let isMutating: Bool
    public let isAddingAccount: Bool
    public let activeIdentityConfirmed: Bool
    public let error: String?
    public let strings: [String: String]
}

extension AccountController {
    public var snapshot: AccountSnapshot {
        AccountSnapshot(
            accounts: accounts.map { account in
                AccountSnapshot.Row(profile: account, initials: account.initials,
                    usage: usageStates[account.id]?.displayedUsage,
                    usageError: usageStates[account.id]?.presentationError,
                    usageStatus: usageStates[account.id]?.presentationStatus ?? "idle",
                    isActive: isAccountActive(account), isCredentialOwner: isCredentialOwner(account),
                    canRemove: canRemoveAccount(account))
            },
            providers: providers.map { provider in
                AccountSnapshot.ProviderRow(profile: provider, isActive: isProviderActive(provider),
                    subtitle: providerSubtitle(provider))
            },
            activeProviderID: activeProviderID,
            authenticationKind: activeAuthentication.map {
                switch $0 { case .signedOut: "signedOut"; case .apiKey: "apiKey"; case .chatGPT: "chatgpt" }
            } ?? "unknown",
            pendingSwitch: pendingSwitch,
            managedProviders: managedProviders,
            providerEditor: providerEditor,
            activeAccountID: accounts.first(where: { isAccountActive($0) })?.id,
            settings: settings,
            isMutating: isMutating, isAddingAccount: isAddingAccount,
            activeIdentityConfirmed: activeIdentityConfirmed,
            error: visibleError.map { $0.messageKey.map(text) ?? $0.message },
            strings: L10n.allStrings(language: settings.language)
        )
    }
}

private extension UsageViewState {
    var presentationStatus: String {
        switch self { case .idle: "idle"; case .loaded: "loaded"; case .stale: "stale"; case .unavailable: "unavailable" }
    }
    var presentationError: String? {
        switch self { case let .stale(_, error), let .unavailable(error): error; default: nil }
    }
}
